js_intermediate/08js_higher_order_functions.md

```diff
@@ -0,0 +1,219 @@
+
+var arr = [5, 3, 2, 9, 8]
+
+for(let i=0; i < arr.length; i++){
+    console.log(arr[i] * 3);
+}
+
+```
+
+Now let's look at this same objective with .map()
+
+```JavaScript
+
+var arr = [5, 3, 2, 9, 8]
+
+var newArr = arr.map(function(index){
+    return index * 3
+})
+
+console.log(newArr);
+
+```
+
+And with the new ES6 syntax we can reduce the code even further...
+
+```JavaScript
+
+var arr = [5, 3, 2, 9, 8]
+
+var newArr = arr.map(value => value * 3)
+
+console.log(newArr);
+
+```
+
+And finally let's put all that together in a function that can take in any array.
+
+#### Before .map() and ES6 Syntax
+
+```JavaScript
+
+var arr = [5, 3, 2, 9, 8]
+
+function mult3(array){
+    let newArr = []
+    for(let i = 0; i < array.length; i++){
+        newArr.push(array[i] * 3)
+    }
+    return newArr
+}
+
+console.log(mult3(arr));
+```
+Notice with the above example that we are iterating through an array of numbers, multiplying each number by three, pushing each number into a new array and returning that new array.  
+
+
+#### Using .map() and ES6 Syntax
+
+```JavaScript
+
+var arr = [5, 3, 2, 9, 8]
+
+    mult3 = (array) => {
+        let newArr = array.map(value => value * 3)
+        return newArr
+    }
+
+console.log(mult3(arr));
+
+```
+Notice that .map() loops through the specified array AND pushes the updated values into a new array.
+
+
+## .filter()
+
+The .filter() function loops through an array and returns a new array with the values that are *truthy* or the values that satisfy your stated condition.
+
+Let's look at a .filter() example by refactoring a function from a previous lesson.
+
+
+#### Before .filter() and ES6 Syntax
+
+
+```JavaScript
+
+var arr1 = [1, 2, 7, 4, 10, 8, 9]
+
+function only_even(array){
+    let newArr = []
+    for(let i = 0; i < array.length; i++){
+        if(array[i] % 2 === 0){
+            newArr.push(array[i])
+        }
+    }
+    return newArr
+}
+
+console.log(only_even(arr1));
+```
+
+So in the above example we have created a function that takes an array as an argument, loops through the array and returns a new array containing only the even numbers.
+
+#### Using .filter() and ES6 Syntax
+
+```JavaScript
+
+var arr = [1, 2, 7, 4, 10, 8, 9]
+
+    only_even = (array) => {
+        let newArr = array.filter(value => value % 2 === 0)
+        return newArr
+    }
+
+console.log(only_even(arr));
+
+```
+
+## Objects
+
+You can use .map() or .filter() on arrays of objects as well.  
+
+Example:  Consider this toons variable.  Use .map() or .filter() to return a new array with only the dogs.
+
+```JavaScript
+
+var toons = [{name: "Brian", animal: "dog"},
+            {name: "Tom", animal: "duck"},
+            {name: "Bugs", animal: "dog"},
+            {name: "Daffy", animal: "duck"},
+            {name: "Felix", animal: "dog"},
+            {name: "Tabby", animal: "cat"}]
+
+
+var animal = toons.filter(value => value.animal === "dog")
+
+
+
+console.log(animal);
+
+
+            ```
```

Comment:
Looks like there's a missing markdown tag here. Check view to make sure there aren't any unwanted effects.